distributed/stealing.py

```diff
@@ -415,4 +425,42 @@ def story(self, *keys):
         return out
 
 
+def _has_restrictions(ts):
+    """Determine whether the given task has restrictions and whether these
+    restrictions are strict.
+    """
+    return not ts.loose_restrictions and (
+        ts.host_restrictions or ts.worker_restrictions or ts.resource_restrictions
+    )
```

Comment:
I wonder if maybe we can cache this on the `TaskState` object itself in order to avoid recomputation.

```python
if ts.restricted:
    ...
```

Reply:
It seems that sometimes these are updated

```python
# in scheduler.py L1707
                    ts = self.tasks[k]
                    ts.loose_restrictions = True
```

If that's the case, I'd prefer to not cache, so that we don't have to worry about invalidating the cache (though perhaps this is the only case?).

Reply:
That's in `update_graph` which is the time when we construct most tasks. My guess is that that this isn't an issue.  We won't check `ts.restricted` before calling those lines.